File: Daily/1137.n-th-tribonacci-number.js
/*
 * @lc app=leetcode id=1137 lang=javascript
 *
 * [1137] N-th Tribonacci Number
 */

// @lc code=start
/**
 * @param {number} n
 * @return {number}
 */
var tribonacci = function (n) {
  /** @type {Uint32Array} */
  const last3Numbers = new Uint32Array([0, 1, 1]);
  if (n < 3) return last3Numbers[n];
  for (let i = 3; i <= n; i++) {
    [last3Numbers[0], last3Numbers[1], last3Numbers[2]] = [
      last3Numbers[1],
      last3Numbers[2],
      last3Numbers[0] + last3Numbers[1] + last3Numbers[2],
    ];
  }

  return last3Numbers[2];
};
// @lc code=end

File: Daily/1137.n-th-tribonacci-number.cpp
/*
 * @lc app=leetcode id=1137 lang=cpp
 *
 * [1137] N-th Tribonacci Number
 */

#include <algorithm>
#include <cstring>

// @lc code=start
class Solution {
 public:
  int tribonacci(int n) {
    int dp[3] = {0, 1, 1};

    for (int i = 3; i <= n; i++) {
      dp[i % 3] = dp[(i - 1) % 3] + dp[(i - 2) % 3] + dp[(i - 3) % 3];
    }

    return dp[n % 3];
  }
};
// @lc code=end

int main() {
  Solution sol;
  sol.tribonacci(4);
}